otherlibs/systhreads/st_pthreads.h

```diff
@@ -297,15 +315,27 @@ static int st_event_wait(st_event e)
 static void * caml_thread_tick(void * arg)
 {
   int *domain_id = (int *) arg;
+  struct timeval curtime;
+  struct timespec deadline;
 
   caml_init_domain_self(*domain_id);
   caml_domain_state *domain = Caml_state;
 
-  while(! atomic_load_acquire(&Tick_thread_stop)) {
-    st_msleep(Thread_timeout);
-
+  caml_plat_lock (&Tick_thread_control.mu);
+  while(1){
+    if (Tick_thread_control.state == Tick_stop) break;
+    gettimeofday (&curtime, NULL);
+    deadline.tv_nsec = 1000 * (curtime.tv_usec + 1000 * Thread_timeout);
+    if (deadline.tv_nsec > 1000000000){
+      deadline.tv_nsec -= 1000000000;
+      deadline.tv_sec = curtime.tv_sec + 1;
+    }else{
+      deadline.tv_sec = curtime.tv_sec;
+    }
+    (void) caml_plat_timedwait (&Tick_thread_control.cond, &deadline);
```

Comment:
Correcting my previous comment, I'd see spurious wakeups handled here:
```suggestion
    while(caml_plat_timedwait (&Tick_thread_control.cond, &deadline))
    {
      if (Tick_thread_control.state == Tick_stop) goto …;
      /* In case of spurious wakeup, keep waiting */
    };
```